### ISSUE TITLE
Motor Remote Control

### ISSUE THREAD

**Issue by bytosaur** [alistair23/AutoMower-BLE#81]
Hey @alistair23 and all,

I understand that this request is not really in the interest of the project, but maybe you found something while scanning through the protocol. 

The mower that I am looking at is the 430X and I want to manually drive it around the lawn. Apparently, the application comes with a feature called _**appDrive**_, but is only available if you have the EPOS system setup at home, which I dont have. So I cannot sniff the protocol which is sad.
I tried searching through the compiled code for this feature and it shows up quite quickly, but I have very little understanding of whats in front of me.

Do you have any insights on that feature?

I really appeciate the effort for your hack. Just by reading your article I have learned a lot.

kind regards,
Paul

**Comment by alistair23** [comment]
For refernce, EPOS is the Husqvarna name for RTK.

I don't have any insights unfortunately. I don't have a 430X so I can't test anything.

It's possible that the firmware won't allow you to use `appDrive` unless there is RTK support, at which point there isn't much we can do (unless you can get around that). But it might not be restricted in the firmware, it's hard to know without seeing the packets and digging into the code.

Obviously sniffing the traffic would be a great start, but if that's not an option then reading through the de-compiled code is your best bet. It does take awhile to get your hear around though.

Maybe someone else with a 430X and RTK support could provide some help?

**Comment by bytosaur** [comment]
Thanks for your quick reply!

Do you think it is possible to simulate the decompiled code in Android Studio?
This would allow to easily enable features in the Android App.
However, I also encounter a bunch of errors dunring decompilation. Not sure why they appear.

As a last resort, I could also try to add some custom hardware to physically bypass the drive control, whenever I want to drive it manually...

**Comment by alistair23** [comment]
Remember that the actual protocol has already been reversed. So you hopefully (unless it's something really different) only need to figure out the major and minor commands and the extra data that is sent.

Again, that assumes that you don't require to the mower to detect RTK support first, before they work. But it's worth a shot

**Comment by erikmol** [comment]
I agree that the most easy approach would be to sniff the trafic to have it implemented, but as an alternative approach I can give you some guidance:

If one opens the Automower (and void any warranty) you can physically hook up to the main board of the Automower and add functionallity. This is exploited by the Robonect project (although not being open source) but there is some Husqvarna provided details in the [HRP repo](https://github.com/HusqvarnaResearch/hrp) (for instance, see the .pdf over there). The interface is a Serial connection (115200 baud rate for newer mowers, 9600 baud rate for older mowers) which you can hook up to a Arduino/ESP32 or similar.

The serial connection have a slighly different format though, compared to the protocol used within this BLE project. The specs is outlined in [this code](https://github.com/HusqvarnaResearch/hrp/blob/c6c0e2a5fb7c6d67e624a29c161d8a0f7271b195/am_driver_safe/src/amproto.cpp#L69) so it's "MessageType" and "Data". As there is a section on ["Remote Navigation"](https://github.com/HusqvarnaResearch/hrp/blob/c6c0e2a5fb7c6d67e624a29c161d8a0f7271b195/am_driver_safe/src/amproto.cpp#L161) I belive that you should be able control it in this fashion. There is also some commands documented in [this document](https://gitlab.com/nbrgmn/wall-e-esp8266-automower/-/blob/develop/Robomower_V0.8.pdf?ref_type=heads) (use Google translate if you don't speak german).

As a side note, the protocol used in this BLE library ("Major", "Minor" pair) is called [AMG3_PROTOCOL_EXTENDED](https://github.com/HusqvarnaResearch/hrp/blob/c6c0e2a5fb7c6d67e624a29c161d8a0f7271b195/am_driver_safe/src/hcp/amg3.c#L231) and it should be possible to send/read these commands in the Serial interface as well. I have not been able to fully understand the implementation though.

**Comment by bytosaur** [comment]
wow, this is very very interesting information!
As warranty is not an isssue I will test this feature ASAP. 

will keep you updated :) thanks Erik
